### ISSUE TITLE
aassign of repeated constant to hash gives odd return

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6886]
Migrated from [rt.perl.org#24380](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=24380) (status was 'resolved')

Searchable as RT24380$

**Comment by p5pRT** [comment]
### From @ysth

#### Created by sthoenna@dhx98431.nonet



aassign to a hash has some problems in list context&#8203;:

$ perl5\.8\.2 \-wle'our %x; print join "&#8203;:"\,%x=\(1\)x8'
Use of uninitialized value in join or string at \-e line 1\.
1&#8203;::1&#8203;:1&#8203;:1&#8203;:Use of uninitialized value in join or string at \-e line 1\.
:1&#8203;:1
$ perl5\.8\.2 \-wle'print \[%x=\(1\)x8\]'
Name "main&#8203;::x" used only once&#8203;: possible typo at \-e line 1\.
Bizarre copy of ARRAY in anonlist at \-e line 1\.
$ perl5\.8\.2 \-wle'use Devel&#8203;::Peek; Dump $\_ for %x=\(1\)x8'
Name "main&#8203;::x" used only once&#8203;: possible typo at \-e line 1\.
SV = PVIV\(0x10272038\) at 0x1028608c
&nbsp;  REFCNT = 2
&nbsp;  FLAGS = \(IOK\,POK\,READONLY\,pIOK\,pPOK\)
&nbsp;  IV = 1
&nbsp;  PV = 0x10272898 "1"\\0
&nbsp;  CUR = 1
&nbsp;  LEN = 2
SV = NULL\(0x0\) at 0x10271170
&nbsp;  REFCNT = 2
&nbsp;  FLAGS = \(\)
SV = PVIV\(0x10272038\) at 0x1028608c
&nbsp;  REFCNT = 2
&nbsp;  FLAGS = \(IOK\,POK\,READONLY\,pIOK\,pPOK\)
&nbsp;  IV = 1
&nbsp;  PV = 0x10272898 "1"\\0
&nbsp;  CUR = 1
&nbsp;  LEN = 2
SV = PVIV\(0x10272050\) at 0x10271284
&nbsp;  REFCNT = 2
&nbsp;  FLAGS = \(IOK\,POK\,pIOK\,pPOK\)
&nbsp;  IV = 1
&nbsp;  PV = 0x10272d98 "1"\\0
&nbsp;  CUR = 1
&nbsp;  LEN = 2
SV = PVIV\(0x10272038\) at 0x1028608c
&nbsp;  REFCNT = 2
&nbsp;  FLAGS = \(IOK\,POK\,READONLY\,pIOK\,pPOK\)
&nbsp;  IV = 1
&nbsp;  PV = 0x10272898 "1"\\0
&nbsp;  CUR = 1
&nbsp;  LEN = 2
SV = NULL\(0x0\) at 0x10271170
&nbsp;  REFCNT = 2
&nbsp;  FLAGS = \(\)
SV = PVIV\(0x10272038\) at 0x1028608c
&nbsp;  REFCNT = 2
&nbsp;  FLAGS = \(IOK\,POK\,READONLY\,pIOK\,pPOK\)
&nbsp;  IV = 1
&nbsp;  PV = 0x10272898 "1"\\0
&nbsp;  CUR = 1
&nbsp;  LEN = 2
SV = PVIV\(0x10272050\) at 0x10271284
&nbsp;  REFCNT = 2
&nbsp;  FLAGS = \(IOK\,POK\,pIOK\,pPOK\)
&nbsp;  IV = 1
&nbsp;  PV = 0x10272d98 "1"\\0
&nbsp;  CUR = 1
&nbsp;  LEN = 2

And the really weird&#8203;:

$  perl5\.8\.2 \-wle'use Devel&#8203;::Peek; sub foo \{ Dump \\@&#8203;\_ \} foo\(%x=\(1\)x8\)'
Name "main&#8203;::x" used only once&#8203;: possible typo at \-e line 1\.
SV = RV\(0x10291020\) at 0x10271170
&nbsp;  REFCNT = 1
&nbsp;  FLAGS = \(TEMP\,ROK\)
&nbsp;  RV = 0x10286050
&nbsp;  SV = PVAV\(0x10273750\) at 0x10286050
&nbsp;    REFCNT = 3
&nbsp;    FLAGS = \(\)
&nbsp;    IV = 0
&nbsp;    NV = 0
&nbsp;    ARRAY = 0x1028e448
&nbsp;    FILL = 7
&nbsp;    MAX = 7
&nbsp;    ARYLEN = 0x0
&nbsp;    FLAGS = \(REAL\)
&nbsp;    Elt No\. 0
&nbsp;    SV = PVIV\(0x10272038\) at 0x1029b0d4
&nbsp;      REFCNT = 5
&nbsp;      FLAGS = \(IOK\,POK\,READONLY\,pIOK\,pPOK\)
&nbsp;      IV = 1
&nbsp;      PV = 0x10272898 "1"\\0
&nbsp;      CUR = 1
&nbsp;      LEN = 2
&nbsp;    Elt No\. 1
&nbsp;    SV = RV\(0x10291020\) at 0x10271170
&nbsp;      REFCNT = 1
&nbsp;      FLAGS = \(TEMP\,ROK\)
&nbsp;      RV = 0x10286050
&nbsp;      SV = PVAV\(0x10273750\) at 0x10286050
&nbsp;        REFCNT = 3
&nbsp;        FLAGS = \(\)
&nbsp;        IV = 0
&nbsp;        NV = 0
&nbsp;        ARRAY = 0x1028e448
&nbsp;        FILL = 7
&nbsp;        MAX = 7
&nbsp;        ARYLEN = 0x0
&nbsp;        FLAGS = \(REAL\)
&nbsp;        Elt No\. 0
&nbsp;        SV = PVIV\(0x10272038\) at 0x1029b0d4
&nbsp;          REFCNT = 5
&nbsp;          FLAGS = \(IOK\,POK\,READONLY\,pIOK\,pPOK\)
&nbsp;          IV = 1
&nbsp;          PV = 0x10272898 "1"\\0
&nbsp;          CUR = 1
&nbsp;          LEN = 2
&nbsp;        Elt No\. 1
&nbsp;        SV = RV\(0x10291020\) at 0x10271170
&nbsp;          REFCNT = 1
&nbsp;          FLAGS = \(TEMP\,ROK\)
&nbsp;          RV = 0x10286050
&nbsp;        Elt No\. 2
&nbsp;        SV = PVIV\(0x10272038\) at 0x1029b0d4
&nbsp;          REFCNT = 5
&nbsp;          FLAGS = \(IOK\,POK\,READONLY\,pIOK\,pPOK\)
&nbsp;          IV = 1
&nbsp;          PV = 0x10272898 "1"\\0
&nbsp;          CUR = 1
&nbsp;          LEN = 2
&nbsp;        Elt No\. 3
&nbsp;        SV = PVIV\(0x10272050\) at 0x10271284
&nbsp;          REFCNT = 3
&nbsp;          FLAGS = \(IOK\,POK\,pIOK\,pPOK\)
&nbsp;          IV = 1
&nbsp;          PV = 0x10272d98 "1"\\0
&nbsp;          CUR = 1
&nbsp;          LEN = 2
&nbsp;    Elt No\. 2
&nbsp;    SV = PVIV\(0x10272038\) at 0x1029b0d4
&nbsp;      REFCNT = 5
&nbsp;      FLAGS = \(IOK\,POK\,READONLY\,pIOK\,pPOK\)
&nbsp;      IV = 1
&nbsp;      PV = 0x10272898 "1"\\0
&nbsp;      CUR = 1
&nbsp;      LEN = 2
&nbsp;    Elt No\. 3
&nbsp;    SV = PVIV\(0x10272050\) at 0x10271284
&nbsp;      REFCNT = 3
&nbsp;      FLAGS = \(IOK\,POK\,pIOK\,pPOK\)
&nbsp;      IV = 1
&nbsp;      PV = 0x10272d98 "1"\\0
&nbsp;      CUR = 1
&nbsp;      LEN = 2
Attempt to free unreferenced scalar&#8203;: SV 0x10271170\.
Attempt to free unreferenced scalar&#8203;: SV 0x10271170\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.8.2:

Configured by sthoenna at Thu Oct 30 14:06:28 PST 2003.

Summary of my perl5 (revision 5.0 version 8 subversion 2 patch 21574) configuration:
  Platform:
    osname=cygwin, osvers=1.5.5(0.9432), archname=cygwin-multi-64int
    uname='cygwin_nt-5.1 dhx98431 1.5.5(0.9432) 2003-09-20 16:31 i686 unknown unknown cygwin '
    config_args='-de -Dusedevel -Dmksymlinks -Duse64bitint -Dusemultiplicity -Doptimize=-g -Dman3ext=3pm'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=define use64bitall=undef uselongdouble=undef
    usemymalloc=y, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-DPERL_USE_SAFE_PUTENV -DDEBUGGING -fno-strict-aliasing',
    optimize='-g',
    cppflags='-DPERL_USE_SAFE_PUTENV -DDEBUGGING -fno-strict-aliasing'
    ccversion='', gccversion='3.3.1 (cygming special)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='ld2', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib /lib
    libs=-lgdbm -ldb -lcrypt -lutil -lgdbm_compat
    perllibs=-lcrypt -lutil -lgdbm_compat
    libc=/usr/lib/libc.a, so=dll, useshrplib=true, libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags=' -L/usr/local/lib'

Locally applied patches:
    MAINT21558


@INC for perl v5.8.2:
    /usr/local/lib/perl5/5.8.2/cygwin-multi-64int
    /usr/local/lib/perl5/5.8.2
    /usr/local/lib/perl5/site_perl/5.8.2/cygwin-multi-64int
    /usr/local/lib/perl5/site_perl/5.8.2
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.8.2:
    CYGWIN=tty ntsec title
    HOME=/home/sthoenna
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/local/bin:/usr/bin:/bin:/usr/X11R6/bin:/usr/bin:/cygdrive/c/WINDOWS/system32:/cygdrive/c/WINDOWS:/cygdrive/c/WINDOWS/System32/Wbem:/cygdrive/c/Program Files/Common Files/Adaptec Shared/System:/cygdrive/c/Program Files/Rational/common
    PERL_BADLANG (unset)
    SHELL (unset)

```
</details>

**Comment by p5pRT** [comment]
### From @rgs

Yitzchak Scott\-Thoennes \(via RT\) wrote&#8203;:
> aassign to a hash has some problems in list context&#8203;:
> 
> $ perl5\.8\.2 \-wle'our %x; print join "&#8203;:"\,%x=\(1\)x8'
> Use of uninitialized value in join or string at \-e line 1\.
> 1&#8203;::1&#8203;:1&#8203;:1&#8203;:Use of uninitialized value in join or string at \-e line 1\.
> :1&#8203;:1
> $ perl5\.8\.2 \-wle'print \[%x=\(1\)x8\]'
> Name "main&#8203;::x" used only once&#8203;: possible typo at \-e line 1\.
> Bizarre copy of ARRAY in anonlist at \-e line 1\.

This should be fixed by&#8203;:

Change 21714 by rgs@&#8203;rgs\-home on 2003/11/13 18&#8203;:55&#8203;:37

&nbsp;        Fix bug \[perl \#24380\] : assigning to a hash in list
&nbsp;        or scalar context yielded a wrong value if the list
&nbsp;        contained duplicated keys for the hash\. This is fixed
&nbsp;        by counting the number of duplicate keys and trimming
&nbsp;        the stack by the corresponding number of items\.

Affected files \.\.\.

\.\.\. //depot/perl/pp\_hot\.c\#334 edit
\.\.\. //depot/perl/t/op/hashassign\.t\#2 edit

**Comment by p5pRT** [comment]
@rgs - Status changed from 'new' to 'resolved'
